### PR TITLE
[Core] Restructure core loop for async input preparation 

### LINKED ISSUE #23233: [RFC]: Restructure the core loop to allow more asynchrony
### Motivation.

Today, the core loop looks like:

```python
while True:
    scheduler_output = self.scheduler.schedule()
    model_runner_output = self.model_executor.execute_model(scheduler_output)
    engine_core_outputs = self.scheduler.update_from_output(scheduler_output, model_runner_output)
    yield engine_core_outputs
```

While simple, this structure doesn't allow us to utilize CPU cycles while the GPU is running the model in `execute_model`.

### Proposed Change.

The proposal is to carve out the sampling stage from the `execute_model` method. This way, `execute_model` will be a non-blocking call that returns without no GPU->CPU synchronization at the end. That is:

```python
while True:
    scheduler_output = self.scheduler.schedule()

    # Prepare inputs and execute the model until the last hidden states. This is non-blocking.
    self.model_executor.execute_model(scheduler_output)

    # If structured outputs is used, produce the bitmask here. Otherwise, bitmask is None.
    bitmask = self.scheduler.get_grammar_bitmask(scheduler_output)

    # Prepare sampling metadata and sample the next token ids. This is blocking.
    model_runner_output = self.model_executor.sample(bitmask)

    engine_core_outputs = self.scheduler.update_from_output(scheduler_output, model_runner_output)
    yield engine_core_outputs
```

This gives two performance benefits:
1. (On the worker side) We can overlap the construction of sampling metadata (and logits processor) with `execute_model`.
2. (On the scheduler side) We can overlap the construction of bitmask with `execute_model`.

For async scheduling, the core loop will look like
```python
# Initial step
scheduler_output = self.scheduler.schedule()
self.model_executor.prepare_inputs(scheduler_output)
self.model_executor.execute_model()
bitmask = self.scheduler.get_grammar_bitmask(scheduler_output)
prev_scheduler_output = scheduler_output

while True:
    scheduler_output = self.scheduler.schedule()
    self.model_executor.prepare_inputs(scheduler_output)
    model_output = self.model_executor.sample(bitmask)
    self.model_executor.execute_model()
    bitmask = self.scheduler.get_grammar_bitmask(scheduler_output)
    engine_core_outputs = self.scheduler.update_from_output(prev_scheduler_output, model_output)
    prev_scheduler_output = scheduler_output
```

### Feedback Period.

_No response_

### CC List.

_No response_

### Any Other Things.

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by WoosukKwon**
Closes #23233 

This PR restructures the core loop for both synchronous and asynchronous scheduling.

For sync scheduling, this allows overlapping bitmask construction with the model execution.
For async scheduling, this allows overlapping input preparation (and bitmask) with the model execution. Currently, only deserialization of scheduler outputs is overlapped. This could be expanded in future PRs.
Besides, this PR refactors the async scheduling loop to be easier to understand, and enables structured outputs support.

### CODE REVIEW

File: vllm/v1/executor/abstract.py [hunk 1]
```diff
@@ -80,12 +80,19 @@ def get_kv_cache_specs(self) -> list[dict[str, KVCacheSpec]]:
...
+
+    def sample(
         self,
-        scheduler_output,
+        grammar_bitmask,
+        non_block: bool = True,
     ) -> Union[ModelRunnerOutput, Future[ModelRunnerOutput]]:
-        output = self.collective_rpc("execute_model",
-                                     args=(scheduler_output, ))
+        del non_block
```
Comment (yinghai) [inline vllm/v1/executor/abstract.py:94]: what's this for?

**Comment by Ronald1995** [comment]
I don't understand how you implement the overlap of prepare_input , it seems like the prepare_input, execute_model, sample task are executed in order in worker process. 

https://github.com/vllm-project/vllm/pull/23811 my solution is to use two threads in worker process, and overlap the prepare_input with the d2h copy operations! would you please explain why your solution will overlap the prepare_input. thanks!
